### PR TITLE
[Merged by Bors] - feat(LineDeriv/QuadraticMap): new file

### PR DESCRIPTION
**Pull request by urkud**
Line derivative of a quadratic form is given by its polar bilinear form.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 1]
```diff
@@ -95,6 +95,11 @@ def polar (f : M → R) (x y : M) :=
   f (x + y) - f x - f y
 #align quadratic_form.polar QuadraticForm.polar
 
+theorem map_add_eq_polar (f : M → R) (x y : M) :
+    f (x + y) = f x + f y + polar f x y := by
+  rw [polar]
+  abel
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:101]: This is in #14215 with a different name

File: Mathlib/Analysis/Calculus/LineDeriv/QuadraticForm.lean [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    HasLineDerivAt 𝕜 f (polar f a b) a b := by
+  simpa [HasLineDerivAt, map_add_eq_polar, f.map_smul, mul_assoc]
+    using ((hasDerivAt_const (0 : 𝕜) (f a)).add
+      ((hasDerivAt_id _).mul (hasDerivAt_mul_const _))).add (hasDerivAt_mul_const _)
+
+theorem lineDifferentiableAt (f : QuadraticForm 𝕜 E) (a b : E) : LineDifferentiableAt 𝕜 f a b :=
+  (f.hasLineDerivAt a b).lineDifferentiableAt
+
+@[simp]
+protected theorem lineDeriv (f : QuadraticForm 𝕜 E) : lineDeriv 𝕜 f = polar f := by
```
Comment (eric-wieser) [inline Mathlib/Analysis/Calculus/LineDeriv/QuadraticForm.lean:33]:
```suggestion
protected theorem lineDeriv_eq (f : QuadraticForm 𝕜 E) : lineDeriv 𝕜 f = polar f := by
```

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean (hunk 1 above)
Comment (urkud) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:101]: Do you want me to rename it to `map_add`?

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:101]: Yes please; and mark it `protected`

File: Mathlib/Analysis/Calculus/LineDeriv/QuadraticForm.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/Analysis/Calculus/LineDeriv/QuadraticForm.lean:33]: Fair.

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:101]: (or wait for #14215, which should only have one dependency left shortly...)

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 3]
```diff
@@ -95,6 +95,11 @@ def polar (f : M → R) (x y : M) :=
   f (x + y) - f x - f y
 #align quadratic_form.polar QuadraticForm.polar
 
+theorem map_add_eq_polar (f : M → R) (x y : M) :
```
Comment (urkud) [inline Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:98]:
```suggestion
protected theorem map_add (f : M → R) (x y : M) :
```

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean (hunk 1 above)
Comment (urkud) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean]: I renamed for now.

**Review by eric-wieser — APPROVED** [review]
bors d+

LGTM, though I'd prefer to wait for #7569 to land first so that it doesn't pick up any more conflicts. Maybe this will hold more generally after that too.

**Comment by urkud** [comment]
OK, I'll wait.

**Comment by eric-wieser** [comment]
#7569 is now in. #14215 (the one which adds the same `add` lemma you have here) is also now dependency-free, but I'm happy for this to go in first if you get around to resolving conflicts here before reviewing there.

**Comment by urkud** [comment]
@eric-wieser I resolved conflicts and generalized to `QuadraticMap`.

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**